x-pack/metricbeat/module/azure/monitor/client_helper.go

```diff
@@ -17,15 +22,64 @@ import (
 
 const missingNamespace = "no metric definitions were found for resource %s and namespace %s. Verify if the namespace is spelled correctly or if it is supported by the resource in case."
 
+func getMetricsDefinitionsWithRetry(client *azure.Client, resource *armresources.GenericResourceExpanded, namespace string) (armmonitor.MetricDefinitionCollection, error) {
+	for {
+
+		metricDefinitions, err := client.AzureMonitorService.GetMetricDefinitions(*resource.ID, namespace)
+		if err != nil {
+			errorMsg := "no metric definitions were found for resource " + *resource.ID + " and namespace " + namespace
+
+			var respError *azcore.ResponseError
+			ok := errors.As(err, &respError)
+			if !ok {
+				return metricDefinitions, fmt.Errorf("%s, failed to cast error to azcore.ResponseError", errorMsg)
+			}
+			// Check for TooManyRequests error and retry if it is the case
+			if respError.StatusCode != http.StatusTooManyRequests {
+				return metricDefinitions, fmt.Errorf("%s, %w", errorMsg, err)
+			}
+
+			// Check if the error has the header Retry After.
+			// If it is present, then we should try to make this request again.
+			retryAfter := respError.RawResponse.Header.Get("Retry-After")
+			if retryAfter == "" {
+				return metricDefinitions, fmt.Errorf("%s %w, failed to find Retry-After header", errorMsg, err)
+			}
+
+			duration, errD := time.ParseDuration(retryAfter + "s")
+			if errD != nil {
+				return metricDefinitions, fmt.Errorf("%s, failed to parse duration %s from header retry after", errorMsg, retryAfter)
+			}
+
+			client.Log.Infof("%s, metricbeat will try again after %s seconds", errorMsg, retryAfter)
+			time.Sleep(duration)
+		} else {
+			return metricDefinitions, err
+		}
+	}
```

Comment:
What about moving the retry/after logic inside the AzureMonitorService?

It seems more like an implementation detail that should stay in the service and not concern the `client.AzureMonitorService.GetMetricDefinitions()` users.

Reply:
I moved it to monitor service, is it ok now?